influxdb3_write/src/write_buffer/persister.rs

```diff
@@ -0,0 +1,940 @@
+use arrow::array::TimestampNanosecondArray;
+use arrow::record_batch::RecordBatch;
+use data_types::{
+    ChunkId, ChunkOrder, PartitionKey, TableId, TimestampMinMax, TransitionPartitionId,
+};
+use datafusion_util::stream_from_batches;
+use iox_query::chunk_statistics::{create_chunk_statistics, NoColumnRanges};
+use iox_query::frontend::reorg::ReorgPlanner;
+use iox_query::QueryChunk;
+use iox_time::TimeProvider;
+use observability_deps::tracing::{error, info};
+use parking_lot::RwLock;
+use parquet::format::FileMetaData;
+use schema::sort::SortKey;
+use schema::Schema;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::watch;
+use tokio::time::MissedTickBehavior;
+
+#[cfg(test)]
+const PERSISTER_CHECK_INTERVAL: Duration = Duration::from_millis(10);
+
+#[cfg(not(test))]
+const PERSISTER_CHECK_INTERVAL: Duration = Duration::from_secs(1);
+
+pub(crate) async fn run_buffer_segment_persist_and_cleanup<P, T, W>(
+    persister: Arc<P>,
+    segment_state: Arc<RwLock<SegmentState<T, W>>>,
+    mut shutdown_rx: watch::Receiver<()>,
+    time_provider: Arc<T>,
+    wal: Option<Arc<W>>,
+    executor: Arc<iox_query::exec::Executor>,
+) where
+    P: Persister,
+    persister::Error: From<<P as Persister>::Error>,
+    T: TimeProvider,
+    W: Wal,
+    write_buffer::Error: From<<P as Persister>::Error>,
+{
+    loop {
+        tokio::select! {
+            _ = shutdown_rx.changed() => {
+                break;
+            }
+            _ = tokio::time::sleep(PERSISTER_CHECK_INTERVAL) => {
+                if let Err(e) = persist_and_cleanup_ready_segments(Arc::clone(&persister), Arc::clone(&segment_state), Arc::clone(&time_provider), wal.clone(), Arc::clone(&executor)).await {
+                    error!("Error persisting and cleaning up segments: {}", e);
+                }
+            }
+        }
+    }
+}
+
+async fn persist_and_cleanup_ready_segments<P, T, W>(
+    persister: Arc<P>,
+    segment_state: Arc<RwLock<SegmentState<T, W>>>,
+    time_provider: Arc<T>,
+    wal: Option<Arc<W>>,
+    executor: Arc<iox_query::exec::Executor>,
+) -> Result<(), crate::Error>
+where
+    P: Persister,
+    persister::Error: From<<P as Persister>::Error>,
+    T: TimeProvider,
+    W: Wal,
+    write_buffer::Error: From<<P as Persister>::Error>,
+{
+    // this loop is where persistence happens so if anything is in persisting,
+    // it's either been dropped or remaining from a restart, so clear those out first.
+    let persisting_segments = {
+        let segment_state = segment_state.read();
+        segment_state.persisting_segments()
+    };
+
+    for segment in persisting_segments {
+        persist_closed_segment_and_cleanup(
+            segment,
+            Arc::clone(&persister),
+            Arc::clone(&segment_state),
+            wal.clone(),
+            Arc::clone(&executor),
+        )
+        .await
+        .unwrap()
```

Comment:
This is pre-existing code, but can you `?` propagate the error here instead of unwrapping? Same for when this function is called below.